### PR TITLE
Add error checking to load_definition function

### PR DESCRIPTION
**Pull request by CarolineMorton**
This is a replacement of #609 

This [ticket](#533) involves returning sensible and useful error messages to the user when they type in a command such as `python -m databuilder generate-dataset dataset_definition.py --dsn csvs`. This changes aims to pick up code that is incorrect in the dataset-definition python file -in this case, `dataset_definition.py`. 

We discussed as a team and decided that we wanted a partial traceback that started with the user input and missed out the error being propagated through multiple internal functions with the databuilder architecture. It was not enough however to just take the last 2-3 lines of the traceback however as the error could be within a function defined in the data definition file.

The sorts of error we might see in a dataset-definition file include:

- Variable does not exist

`dataset_definition.py`
```python
asdf
```

- Incorrect dataset construction

`dataset_definition.py`
```python
from databuilder.query_language import Dataset, build_patient_table

table = build_patient_table("table", {"i": int, "b": bool})

dataset = Dataset()
dataset.v = table.i + table.b
```

## What I have done
Thanks to the previous review by @evansd, I have moved the code into `load_definition(). The code then walks the traceback stack until it finds one where the origin is from the dataset code. 

This allows the following message to print to the screen (respectively as above):

```bash
Error Type: <class 'NameError'>
File: dataset_definition.py
Line number 1: name 'asdf' is not defined
```

```bash
Error Type: <class 'ImportError'>
File: dataset_definition.py
Line number 1: cannot import name 'build_patient_table' from 'databuilder.query_language' (/Users/carolinemorton/Documents/Datalab-Github/databuilder/databuilder/query_language.py)
```

### CODE REVIEW

**Review by inglesp — COMMENTED** [review]
It's not clear to me that this addresses the kind of error described in [this comment](https://github.com/opensafely-core/databuilder/pull/609#discussion_r921305369).  Unfortunately, the codebase has changed since the comment was written, and so the code now raises an `ImportError` on line 1, instead of an error from the internals of databuilder.

You'll now want to test with something like:

```
from databuilder.query_language import PatientFrame, Series, construct

@construct
class table(PatientFrame):
    i = Series(int)
    b = Series(bool)

dataset = Dataset()
dataset.v = table.i + table.b
```

Can you try with this example, and share the output?  We'd expect to see the error from the last line (cos you can't add an int to a bool).

**Comment by CarolineMorton** [comment]
You get:
```bash
Error Type: <class 'databuilder.query_model.TypeValidationError'>
File: dataset_definition.py
Line number 1: Add.rhs requires 'databuilder.query_model.Series[int]' but got 'databuilder.query_model.Series[bool]'
In node:
Function.Add(lhs=SelectColumn(source=SelectPatientTable(name='table', schema=TableSchema(i=int, b=bool)), name='i'), rhs=SelectColumn(source=SelectPatientTable(name='table', schema=TableSchema(i=int, b=bool)), name='b'))
```

**Comment by inglesp** [comment]
Where's it getting `Line number 1` from?

**Comment by CarolineMorton** [comment]
Pushed a fix for this. Error message displayed is now:

```bash
Error Type: <class 'databuilder.query_model.TypeValidationError'>
File: dataset_definition.py
Line number 10: Add.rhs requires 'databuilder.query_model.Series[int]' but got 'databuilder.query_model.Series[bool]'
In node:
Function.Add(lhs=SelectColumn(source=SelectPatientTable(name='table', schema=TableSchema(i=int, b=bool)), name='i'), rhs=SelectColumn(source=SelectPatientTable(name='table', schema=TableSchema(i=int, b=bool)), name='b'))
```

**Review by inglesp — CHANGES_REQUESTED** [review]
Have left a few questions on the PR.  This also needs a test.  Would you like to discuss later today?

File: databuilder/main.py [hunk 1]
```diff
@@ -140,11 +140,24 @@ def test_connection(backend_class, url, environ):
...
+    try:
+        module = load_module(definition_file)
+    except Exception:
+        exc_type, exc_value, exc_tb = sys.exc_info()
+        while exc_tb:
+            if exc_tb.tb_frame.f_globals["__name__"] == "dataset":
+                print(f"Error Type: {exc_type}")
+                print(f"File: {exc_tb.tb_frame.f_code.co_filename}")
+                print(f"Line number {exc_tb.tb_lineno}: {exc_value}")
+                sys.exit(2)
```
[the remark this answers is not in the document]
Comment (CarolineMorton) [reply databuilder/main.py]: yes, as discussed.

File: databuilder/main.py [hunk 2]
```diff
@@ -140,11 +140,24 @@ def test_connection(backend_class, url, environ):
...
+            exc_tb = exc_tb.tb_next
+        sys.exit(2)
```
[the remark this answers is not in the document]
Comment (CarolineMorton) [reply databuilder/main.py]: Chnaged to 1

File: databuilder/main.py [hunk 3]
```diff
@@ -140,11 +140,24 @@ def test_connection(backend_class, url, environ):
 
 
 def load_definition(definition_file):
-    module = load_module(definition_file)
+    try:
+        module = load_module(definition_file)
+    except Exception:
+        exc_type, exc_value, exc_tb = sys.exc_info()
+        while exc_tb:
+            if exc_tb.tb_frame.f_globals["__name__"] == "dataset":
```
[the remark this answers is not in the document]
Comment (CarolineMorton) [reply databuilder/main.py:151]: as discussed. I have added comment.

**Comment by inglesp** [comment]
Superseded by #711.

**Pull request closed without merging**